### ISSUE TITLE
issue in deployment of IoT stack

### ISSUE THREAD

**Issue by nikita-b1** [aws-samples/aws-iot-sap-condition-monitoring-demo#1]
Hello,
 
I am trying to deploy an IoT stack, but facing issue while trying to deploy initial stack using following 
command:
cdk deploy iot -O=iot-outputs.json

Error:
`Traceback (most recent call last):
  File "app.py", line 3, in <module>
    from aws_cdk import core
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\core\__init__.py", line 7154, in <module>
    class ConstructNode(metaclass=jsii.JSIIMeta, jsii_type="@aws-cdk/core.ConstructNode"):
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\core\__init__.py", line 7205, in ConstructNode
    runtime_info: typing.Optional[aws_cdk.cx_api.RuntimeInfo] = None,
AttributeError: module 'aws_cdk.cx_api' has no attribute 'RuntimeInfo'`

https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_cx-api.MetadataEntry.html
above link shows, RuntimeInfo and Metadata are deprecated

But being used in this project.

Could you please help me to solve this problem

**Comment by frumania** [comment]
Hi,
have you updated to the latest version of CDK, by chance?
> npm install -g aws-cdk@latest

**Comment by nikita-b1** [comment]
Thanks frumania, I've used this npm command but unfortunately it's not working.
would be appreciated to have another solution regarding it.

**Comment by frumania** [comment]
I just tested the deploment via a AWS Cloud9 instance (for ease of use).
Works fine!
Due to a recent CDK Bug, just make sure to run initially
`npm i -g aws-cdk@latest-1`

**Comment by nikita-b1** [comment]
I updated cdk using this command and it is successful. But when I try to bootstrap using command
`cdk bootstrap aws://<aws-id>/<region>`
  following error occurs

`Traceback (most recent call last):
  File "app.py", line 3, in <module>
    from aws_cdk import core
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\__init__.py", line 24609, in <module>
    from . import aws_apigateway
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_apigateway\__init__.py", line 1549, in <module>
    from ..aws_certificatemanager import ICertificate as _ICertificate_c194c70b
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_certificatemanager\__init__.py", line 184, in <module>
    from ..aws_cloudwatch import (
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_cloudwatch\__init__.py", line 549, in <module>
    from ..aws_iam import Grant as _Grant_a7ae64f8, IGrantable as _IGrantable_71c4f5de
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_iam\__init__.py", line 662, in <module>
    class AddToPrincipalPolicyResult:
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_iam\__init__.py", line 667, in AddToPrincipalPolicyResult
    policy_dependable: typing.Optional[constructs.IDependable] = None,
AttributeError: module 'constructs' has no attribute 'IDependable'

NOTICES

19836   AWS CDK v1 entering maintenance mode soon

        Overview: AWS CDK v1 is entering maintenance mode on June 1, 2022.
                  Migrate to AWS CDK v2 to continue to get the latest features
                  and fixes!

        Affected versions: framework: 1.*, cli: 1.*

        More information at: https://github.com/aws/aws-cdk/issues/19836


If you don’t want to see a notice anymore, use "cdk acknowledge <id>". For example, "cdk acknowledge 19836".
Subprocess exited with error 1`

**Comment by frumania** [comment]
I see.
Then run 
`cdk acknowledge 19836`
and
`cdk bootstrap aws://<aws-id>/<region>`
but replace by **your** AWSAccountID and Target AWS Region
e.g.
`cdk bootstrap aws://529824580565/us-east-1`

**Comment by nikita-b1** [comment]
Tried, but again same error,

`Traceback (most recent call last):
  File "app.py", line 3, in <module>
    from aws_cdk import core
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\__init__.py", line 24609, in <module>
    from . import aws_apigateway
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_apigateway\__init__.py", line 1549, in <module>
    from ..aws_certificatemanager import ICertificate as _ICertificate_c194c70b
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_certificatemanager\__init__.py", line 184, in <module>
    from ..aws_cloudwatch import (
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_cloudwatch\__init__.py", line 549, in <module>
    from ..aws_iam import Grant as _Grant_a7ae64f8, IGrantable as _IGrantable_71c4f5de
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_iam\__init__.py", line 662, in <module>
    class AddToPrincipalPolicyResult:
  File "C:\aws-iot-sap-condition-monitoring-demo\cdk-iot-analytics\.venv\lib\site-packages\aws_cdk\aws_iam\__init__.py", line 667, in AddToPrincipalPolicyResult
    policy_dependable: typing.Optional[constructs.IDependable] = None,
AttributeError: module 'constructs' has no attribute 'IDependable'`

**Comment by frumania** [comment]
Ok, just to crosscheck:

1) You are using AWS Cloud 9
2) Local Env. Setup -> https://github.com/aws-samples/aws-iot-sap-condition-monitoring-demo#set-up-local-environment
3) Maintained cdk.json ([example](https://catalog.workshops.aws/sap-on-aws-beyond/en-US/immersion-days/data-and-analytics/lab4))

**Comment by nikita-b1** [comment]
1.we are using local system, not AWS cloud 9
2.Local Env. setup is done as mentioned
3.Yes, cdk.json is maintained
(just to confirm, in CDK.json 
    "sapUsername": "<user>",
    "sapPassword": "<password>",
here, we use S-ID or sap system ID?

**Comment by frumania** [comment]
Ok, i suggest to use Cloud 9, just for deployment.
Local Env. can have so many side effects :(

"sapUsername": "",
"sapPassword": "",

would actually be your **SAP Backend User & Password**.

Marcel

**Comment by nikita-b1** [comment]
Thank you Frumania for your help.

I have started deployment on AWS CLoud9.

Error occured in updated Detector model to latest version using command:
`aws iotevents update-detector-model --cli-input-json file://cdk_sap_blog/analytics/detector_model.json`

Error:
`An error occurred (InvalidRequestException) when calling the UpdateDetectorModel operation: Unable to assume role, role ARN: arn:aws:iam::076121412771:role/CDKSAPBlogEventDetectorRole`

**Comment by nikita-b1** [comment]
Thank You Frumania,

I have successfully deployed it. I can see temperature state in detector model but did not receive any mail, even if that over/under temperature lasts for more than 1-2 mins.

**Comment by frumania** [comment]
Hi,
if stack is deployed and the SAP system is configured correctly, you can follow the related workshop content:
https://catalog.workshops.aws/sap-on-aws-beyond/en-US/immersion-days/data-and-analytics/lab4

In case DetectorModel shows "Overtemp", but you do not get any notification, check the CloudWatch Logs for the following Lambda Function to troubleshoot:
**CDK-SAP-Blog-OData-Function**

**Comment by nikita-b1** [comment]
Hello Frumania,

I checked the log,
it is showing this error
`START RequestId: 2b3de090-eaf0-4e23-b672-490e4137325a Version: $LATEST
event: {'key1': 'value1', 'key2': 'value2', 'key3': 'value3'}
Connection to SAP System timed out: SSLError
Is this the correct hostname and port number?
END RequestId: 2b3de090-eaf0-4e23-b672-490e4137325a
REPORT RequestId: 2b3de090-eaf0-4e23-b672-490e4137325a	Duration: 2209.09 ms	Billed Duration: 2210 ms	Memory Size: 128 MB	Max Memory Used: 72 MB	Init Duration: 486.95 ms	
RequestId: 2b3de090-eaf0-4e23-b672-490e4137325a Error: Runtime exited with error: exit status 1
Runtime.ExitError`

I rechecked host name and port number and it is working fine with browser.
but not able to access it from AWS.

Is there anything to configure or any additional settings need to make?
